src/util/fs.cc

```diff
@@ -121,4 +129,460 @@ bool os_portable_copy_file_to_dir(const std::string& src,
+      if (rdt & redirect_t::STDERR)
+        dup2(fd, STDERR_FILENO);
+      close(fd);
+    }
+    
+    // this is memory leak, but I have no other way...
+    // hope this will be reclaimed by OS
+    // + 1 for NULL
+
+    const int MAX_ARG = 64;
+    char * argv[MAX_ARG];
+    if (cmdargs.size() >= MAX_ARG) {
+      ILA_ERROR <<"Too many args";
+      report_to_parent = execute_result::ARG;
+      write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+      exit(1);
+    }
+    //ILA_ASSERT(cmdargs.size() < MAX_ARG) << "Too many args";
+
+    for(auto it = cmdargs.begin(); it != cmdargs.end(); ++ it) {
+      // this is memory leak, I know, but what can I do ?
+      argv[it-cmdargs.begin()] = new char[it->size() + 1];
+      strcpy(argv[it-cmdargs.begin()], it->c_str());
+    }
+    argv[ cmdargs.size() ] = NULL;
+
+    report_to_parent = execute_result::NONE;
+    write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+
+    int ret = execvp(cmdargs[0].c_str(), argv);
+    // if not successful
+    if (ret == -1) {
+      report_to_parent = execute_result::NONE;
+      write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+    }
+    close(pipefd[1]);
+    exit(ret);
+  } else {
+    // The parent will wait for its end
+    int infop;
+    unsigned child_report;
+    struct sigaction new_act;
+    struct sigaction old_act;
+
+    close(pipefd[1]); // close the write end
+    read(pipefd[0], (void *) &child_report, sizeof(child_report));
+    if (child_report != execute_result::NONE) {
+      // the child has error before exec
+      _ret.failure = static_cast<execute_result::_failure>( child_report) ;
+      close(pipefd[0]);
+      return _ret;
+    }
+
+    child_pid = pid;
+    
+    if (!pid_file_name.empty()) {
+      std::ofstream fout(pid_file_name);
+      fout << child_pid << std::endl;
+    }
+    // set alarm
+    if(timeout != 0) {
+      shared_time_out = 0;
+      new_act.sa_handler = parent_alarm_handler;
+      sigemptyset (&new_act.sa_mask);
+      new_act.sa_flags = 0;
+
+      sigaction(SIGALRM, &new_act, &old_act);
+      alarm(timeout);
+    }
+    // wait for sub-process
+    int wait_pid_res = waitpid(pid, &infop, 0);
+    
+    gettimeofday(&Time2, NULL);
+    _ret.seconds = 
+      ( 
+        (Time2.tv_usec + Time2.tv_sec*1000000.0) -  
+        (Time1.tv_usec + Time1.tv_sec*1000000.0) ) / 1000000.0;
+
+    if(timeout != 0) {
+      alarm(0); //cancel if previously set
+      sigaction(SIGALRM, &old_act, NULL); // restore the old one
+    }
+    
+    if (timeout == 0 && wait_pid_res == -1) {
+      _ret.failure = execute_result::WAIT;
+      perror(NULL);
+      close(pipefd[0]);
+      return _ret;
+    }
+
+    if (wait_pid_res != -1)
+      _ret.subexit_normal = WIFEXITED(infop);
+
+    // read again, if exec suceeded, it should be EOF (read will fail)
+    int sec_read = read(pipefd[0], (void *) &child_report, sizeof(child_report));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)